src/platform/cmx/boot_comp.c

```diff
@@ -0,0 +1,279 @@
+////		u8_t   *p  = ptes + i * PAGE_SIZE;
+////		paddr_t pf = chal_va2pa(p);
+////
+////		pgtbl_init_pte(p);
+////		//pte_cap->pgtbl = (pgtbl_t)p;
+////
+////		/* hook the pte into the boot component's page tables */
+////		ret = cap_cons(ct, pgdcap, ptecap, (capid_t)(user_vaddr + i*PGD_RANGE));
+////		assert(!ret);
+////	}
+////
+////	printk("\tMapping in %s.\n", label);
+////	/* Map in the actual memory. */
+////	for (i = 0 ; i < range/PAGE_SIZE ; i++) {
+////		u8_t *p     = kern_vaddr + i * PAGE_SIZE;
+////		paddr_t pf  = chal_va2pa(p);
+////		u32_t mapat = (u32_t)user_vaddr + i * PAGE_SIZE, flags = 0;
+////
+////		if (uvm  && pgtbl_mapping_add(pgtbl, mapat, pf, PGTBL_USER_DEF))
+////			assert(0);
+////		if (!uvm && pgtbl_cosframe_add(pgtbl, mapat, pf, PGTBL_COSFRAME))
+////			assert(0);
+////		assert((void*)p == pgtbl_lkup(pgtbl, user_vaddr+i*PAGE_SIZE, &flags));
+////	}
+//
+//	return 0;
+//}
+
+/* FIXME:  loops to create threads/tcaps/rcv caps per core. */
+static void
+kern_boot_thd(struct captbl *ct, void *thd_mem, void *tcap_mem)
+{
+	struct cos_cpu_local_info *cos_info = cos_cpu_local_info();
+	struct thread *t = thd_mem;
+	struct tcap   *tc = tcap_mem;
+	tcap_res_t     expended;
+	int            ret;
+	struct cap_pgtbl *cap_pt;
+	pgtbl_t           pgtbl;
+
+	assert(sizeof(struct cos_cpu_local_info) == STK_INFO_SZ);
+	memset(cos_info, 0, sizeof(struct cos_cpu_local_info));
+	cos_info->cpuid          = 0;
+	cos_info->invstk_top     = 0;
+	cos_info->overflow_check = 0xDEADBEEF;
+
+	ret = thd_activate(ct, BOOT_CAPTBL_SELF_CT, BOOT_CAPTBL_SELF_INITTHD_BASE,
+			   thd_mem, BOOT_CAPTBL_SELF_COMP, 0);
+	assert(!ret);
+
+	tcap_active_init(cos_info);
+	ret = tcap_activate(ct, BOOT_CAPTBL_SELF_CT, BOOT_CAPTBL_SELF_INITTCAP_BASE, tcap_mem, TCAP_PRIO_MAX);
+	assert(!ret);
+	tc->budget.cycles = TCAP_RES_INF; /* Chronos's got all the time in the world */
+	tc->perm_prio     = 0;
+	tcap_setprio(tc, 0);              /* Chronos gets preempted by no one! */
+	list_enqueue(&cos_info->tcaps, &tc->active_list); /* Chronos on the TCap active list */
+	cos_info->tcap_uid  = 1;
+	cos_info->cycles    = tsc();
+	cos_info->curr_tcap = tc;
+
+	thd_current_update(t, t, cos_info);
+
+	ret = arcv_activate(ct, BOOT_CAPTBL_SELF_CT, BOOT_CAPTBL_SELF_INITRCV_BASE,
+			    BOOT_CAPTBL_SELF_COMP, BOOT_CAPTBL_SELF_INITTHD_BASE,
+			    BOOT_CAPTBL_SELF_INITTCAP_BASE, 0, 1);
+	assert(!ret);
+
+	/*
+	 * boot component's mapped into SELF_PT,
+	 * switching to boot component's pgd
+	 */
+	cap_pt = (struct cap_pgtbl *)captbl_lkup(ct, BOOT_CAPTBL_SELF_PT_META);
+	if (!cap_pt || !CAP_TYPECHK(cap_pt, CAP_PGTBL)) assert(0);
+	/* PRY: modified to pass compilation */
+	//pgtbl = cap_pt->pgtbl;
+	assert(pgtbl);
+	pgtbl_update(pgtbl);
+
+	printk("\tCreating initial threads, tcaps, and rcv end-points in boot-component.\n");
+}
+/* PRY: the booing allocations have changed a lot. Not as usual, we do not allocate pages for CM3 */
+void
+kern_boot_comp(void)
```

Comment:
This is not that generic across architectures. Because the page table construction is not only platform specific but also (could-be) application tailored. But we can have a bold outline in the kernel that calls these functions actually.